Fix physics exact match LaTeX answer extraction and update prompt format

Ryan0v0 opened the pull request:
This PR fixes critical issues in the physics exact match answer extraction system that caused correct answers to be incorrectly scored as wrong.

Ryan0v0 committed:
Fix physics exact match LaTeX answer extraction and update prompt format

- Enhanced support for multiple LaTeX formats: \boxed{}, \\boxed{}, \(\boxed{}\)
- Support for \text{Final Answer: } \boxed{} format
- Strict enforcement of 'Final Answer: \boxed{...}' format requirement
- Fixed nested list handling in math_verify_score function
- Updated prompt in _default_exact_match_yaml to use 'Final Answer: \boxed{}' format
- Improved extraction success rate from 94.1% to 97.1%

This addresses LaTeX-wrapped answers not being properly extracted,
leading to false negatives in evaluation results.

HelloJocelynLu commented:
Do you have test results to share? On my side, almost all answers failed to be extracted.
The prompt in lm_eval/tasks/science/basic_knowledge/physics/_default_exact_match.yaml is overwritten by downstream exact-match tasks; the actual prompt comes from the corresponding *_exact_match.yaml files.
If we want to change the answer pattern from <math>xxx</math> to \boxed{xxx}, we need to update those *_exact_match.yaml files instead of the default file.

HelloJocelynLu committed:
update all physics exact match prompts

Ryan0v0 committed:
Fix math_verify_score nested list handling and improve exact match prompts

- Fix critical bug in math_verify_score function where nested lists like [['0']] were not properly extracted
- Add extract_final_value() function to recursively extract final non-list values
- Add normalize_for_comparison() function for proper string comparison
- Update all exact match YAML files to use consistent detailed doc_to_text format
- Improve prompt instructions for better model output formatting

This fix resolves the issue where correct answers were getting 0.0 scores due to improper nested list handling.

Ryan0v0 commented:
The yaml files have been updated accordingly.

Command:
```python
python -m lm_eval --model openai-chat-completions --model_args model=gpt-4o-mini --tasks exact_match_quantum_information,exact_match_astrophysics_cosmology,exact_match_condensed_matter_physics,exact_match_probability_statistics --apply_chat_template --output results/all_exact_match_physics_results_fixed.json --log_samples
```
Result:
<img width="971" height="194" alt="image" src="https://github.com/user-attachments/assets/2caa3185-2ab6-467e-8030-683681992819" />

HelloJocelynLu committed:
Required installation of math_verify for proper comparison

HelloJocelynLu commented:
LGTM, my test result
```
lm_eval --model anthropic-chat-completions --model_args model=claude-3-5-sonnet-20240620 --tasks physics_exact_match --apply_chat_template --output results/phys_sample.json --log_samples
```
<img width="1788" height="258" alt="acd4c7d111afac9dea0025bd3b8fe7f5" src="https://github.com/user-attachments/assets/5944177c-c1a0-4eb8-9257-3abe800720ef" />
